src/adr/0022-cassandra-message-inconsistency.md

```diff
@@ -0,0 +1,88 @@
+
+ - Adding a message:
+   - (CassandraMessageMapper) First reference the message in `messageIdTable` then in `imapUidTable`.
+   - (CassandraMessageIdMapper) First reference the message in `imapUidTable` then in `messageIdTable`.
+ - Deleting a message:
+   - (CassandraMessageMapper) First delete the message in `imapUidTable` then in `messageIdTable`.
+   - (CassandraMessageIdMapper) Read the message metadata using `imapUidTable`, then first delete the message in 
+   `imapUidTable` then in `messageIdTable`.
+ - Copying a message:
+   - (CassandraMessageMapper) Read the message first, then first reference the message in `messageIdTable` then
+    in `imapUidTable`.
+ - Moving a message:
+   - (CassandraMessageMapper) Logically copy then delete. A failure in the chain migh lead to duplicated message (present 
+   in both source and destination mailbox) as well as different view in IMAP/JMAP.
+   - (CassandraMessageIdMapper) First reference the message in `imapUidTable` then in `messageIdTable`.
+ - Updating a message flags:
+   - (CassandraMessageMapper) First update conditionally the message in `imapUidTable` then in `messageIdTable`.
+   - (CassandraMessageIdMapper) First update conditionally the message in `imapUidTable` then in `messageIdTable`.
+
+## Decision
+
+Adopt `imapUidTable` as a source of truth. All writes will be performed to `messageIdTable` then performed on
+`imapUidTable` if successful.
+
+We thus need to modify CassandraMessageMapper 'add' + 'copy' to first write to the source of truth (`imapUidTable`)
+
+We can adopt a retry policy of the `messageIdTable` projection update as a mitigation strategy.
+
+Using `imapUidTable` table as a source of truth, we can rebuild the `messageIdTable` projection:
+ - Iterating `imapUidTable` entries, we can rewrite entries in `messageIdTable`
+ - Iterating `messageIdTable` we can remove entries not referenced in `imapUidTable`
+ - Adding a delay and a re-check before the actual fix can decrease the occurrence of concurrency issues
+
+We will expose a webAdmin task for doing this.
+
+## Consequences
+
+User actions concurrent to the inconsistency fixing task could result in concurrency issues. New inconsistencies could be
+created. However table of truth would not be impacted hence rerunning the inconsistency fixing task will eventually fix 
+all issues.
+
+This task could be run safely online and can be scheduled on a recurring basis outside of peak traffic by an admin to
+ensure Cassandra message consistency.
+
+## References
+
+* [Plan for fixing Cassandra ACL inconsistencies](https://github.com/linagora/james-project/pull/3125)
+
+* [General mailing list discussion about inconsistencies](https://www.mail-archive.com/server-dev@james.apache.org/msg64432.html)
+
+* [Pull Request: JAMES-3058 Concurrency testing for fixing Cassandra mailbox inconsistencies](https://github.com/linagora/james-project/pull/3130)
+
+The delay strategy to decrease concurrency issue occurrence is described here.
```

Comment:
line to be removed?

Reply:
If we are online this line makes sense no?

Reply:
not for me, maybe I missed something. But in this case you miss a link

Reply:
oh, by "here" you mean in this adr?

Reply:
I mean "in the above linked PR". Described in comments, ML exchanges, implemented and tested.